scriptworker/test/test_worker.py

```diff
@@ -254,3 +255,296 @@ def _mocker_run_tasks_helper(mocker, exc, func_to_raise):
+    mocker.patch.object(asyncio, 'sleep', create_async())
+    mocker.patch('scriptworker.worker.prepare_to_run_task', create_async())
+    mocker.patch('scriptworker.worker.reclaim_task', create_async())
+    mocker.patch('scriptworker.worker.do_run_task', create_async(0))
+    mocker.patch('scriptworker.worker.do_upload', create_async(0))
+    mocker.patch('scriptworker.worker.cleanup', create_sync())
+    mock_complete_task = mocker.patch('scriptworker.worker.complete_task')
+
+    future = asyncio.Future()
+    future.set_result(None)
+    mock_complete_task.return_value = future
+
+    run_tasks = RunTasks()
+    await run_tasks.invoke(context)
+    mock_complete_task.assert_called_once_with(mock.ANY, 0)
+
+
+@pytest.mark.asyncio
+async def test_run_tasks_cancel_claim_work(context, mocker):
+    slow_function_called, slow_function = create_slow_async()
+    mocker.patch('scriptworker.worker.claim_work', slow_function)
+
+    mock_sleep = mocker.patch.object(asyncio, 'sleep')
+    mock_sleep.return_value = create_finished_promise()
+
+    mock_prepare_task = mocker.patch('scriptworker.worker.prepare_to_run_task')
+    mock_prepare_task.return_value = create_finished_promise()
+
+    mock_complete_task = mocker.patch('scriptworker.worker.complete_task')
+    mock_complete_task.return_value = create_finished_promise()
+
+    run_tasks = RunTasks()
+    run_tasks_future = asyncio.get_running_loop().create_task(run_tasks.invoke(context))
+    await slow_function_called
+    await run_tasks.cancel()
+    await run_tasks_future
+
+    mock_sleep.assert_not_called()
+    mock_prepare_task.assert_not_called()
+    mock_complete_task.assert_not_called()
+
+
+@pytest.mark.asyncio
+async def test_run_tasks_cancel_sleep(context, mocker):
+    slow_function_called, slow_function = create_slow_async()
+    mocker.patch.object(asyncio, 'sleep', slow_function)
+
+    mocker.patch('scriptworker.worker.claim_work', create_async(_MOCK_CLAIM_WORK_NONE_RETURN))
+
+    mock_prepare_task = mocker.patch('scriptworker.worker.prepare_to_run_task')
+    mock_prepare_task.return_value = create_finished_promise()
+
+    mock_complete_task = mocker.patch('scriptworker.worker.complete_task')
+    mock_complete_task.return_value = create_finished_promise()
+
+    run_tasks = RunTasks()
+    run_tasks_future = asyncio.get_running_loop().create_task(run_tasks.invoke(context))
+    await slow_function_called
+    await run_tasks.cancel()
+    await run_tasks_future
+
+    mock_prepare_task.assert_not_called()
+    mock_complete_task.assert_not_called()
+
+
+@pytest.mark.asyncio
+async def test_run_tasks_cancel_cot(context, mocker):
+    context.config['verify_chain_of_trust'] = True
+
+    slow_function_called, slow_function = create_slow_async()
+    mocker.patch('scriptworker.worker.verify_chain_of_trust', slow_function)
+
+    mock_prepare_task = mocker.patch('scriptworker.worker.prepare_to_run_task')
```

Comment:
These functions are pretty core to the worker, so as mentioned below? even clean arch can't save us from lots of i/o in the core functions. We may be able to mitigate this with kwargs as mentioned in chat.

It may be worthwhile adding some integration tests around cancelling tasks. We already have `test_cancel_task`, but we could verify uploads happen as expected etc.

Reply:
True that these functions are pretty core! Sometimes I just need to complain after setting up a bunch of mocks to let off steam :wink:
I'll update the cot-cancellation and task-cancellation tests such that they verify that `do_upload` is called, too :)
I'll add an integration test as well, good call

Reply:
Integration test added!